src/persist-client/src/internal/trace.rs

```diff
@@ -103,10 +112,223 @@ impl<T: Timestamp + Lattice> Default for Trace<T> {
+                    id_batch.id,
+                    ThinSpineBatch {
+                        level,
+                        desc: id_batch.batch.desc.clone(),
+                        parts: vec![push_hollow_batch(id_batch)],
+                    },
+                ),
+                SpineBatch::Fueled {
+                    id,
+                    desc,
+                    parts,
+                    len: _,
+                } => (
+                    *id,
+                    ThinSpineBatch {
+                        level,
+                        desc: desc.clone(),
+                        parts: parts.into_iter().map(&mut push_hollow_batch).collect(),
+                    },
+                ),
+            };
+            spine_batches.insert(id, spine_batch);
+        };
+
+        for (level, state) in self.spine.merging.iter().enumerate() {
+            match state {
+                MergeState::Vacant => {}
+                MergeState::Single(batch) => push_spine_batch(level, batch),
+                MergeState::Double(left, right, merge) => {
+                    push_spine_batch(level, left);
+                    push_spine_batch(level, right);
+                    fueling_merges.insert(level, merge.clone());
+                }
+            }
+        }
+
+        if !self.roundtrip_structure {
+            assert!(hollow_batches.is_empty());
+            spine_batches.clear();
+            fueling_merges.clear();
+        }
+
+        FlatTrace {
+            since,
+            legacy_batches,
+            hollow_batches,
+            spine_batches,
+            fueling_merges,
+        }
+    }
+    pub(crate) fn unflatten(value: FlatTrace<T>) -> Result<Self, String> {
+        let FlatTrace {
+            since,
+            legacy_batches,
+            mut hollow_batches,
+            spine_batches,
+            mut fueling_merges,
+        } = value;
+
+        // If the flattened representation has spine batches, we know to preserve the structure for
+        // this trace.
+        // Note that for empty spines, roundtrip_structure will default to false. This is done for
+        // backwards-compatability.
+        let roundtrip_structure = !spine_batches.is_empty();
+
+        // We need to look up legacy batches somehow, but we don't have a spine id for them.
+        // Instead, we rely on the fact that the spine must store them in antichain order.
+        let compare_chains = |left: &Antichain<T>, right: &Antichain<T>| {
+            if PartialOrder::less_than(left, right) {
+                Ordering::Less
+            } else if PartialOrder::less_than(right, left) {
+                Ordering::Greater
+            } else {
+                Ordering::Equal
+            }
+        };
+        let mut legacy_batches: Vec<_> = legacy_batches.into_iter().map(|(k, _)| k).collect();
+        legacy_batches.sort_by(|a, b| compare_chains(a.desc.lower(), b.desc.lower()).reverse());
```

Comment:
alternate hack: `|a, b| a.desc.lower().elements().cmp(b.desc.lower().elements())` 🙈

Reply:
It was a bit of a joke, but I think I do actually prefer this alternate. The implementation of `compare_chains` is somewhat alarming to read 😅

Reply:
It's tempting, but I've been having trouble convincing myself that they're equivalent! For example `[] < [0]` for array comparison but the opposite for antichains; we don't have to worry about that specific case for lowers, but I haven't worked through the details for partially-ordered times. (Whereas `compare_chains` is a bit gross, but seemed a bit less fraught from this perspective.)